chat/README.md

```diff
@@ -0,0 +1,21 @@
+# Group Chat
+
+## Overview
+
+This example demonstrates how to implement a group chat with dynamic groups in DAML. All members of a `ChatGroup` have equal rights -- there is no owner, admin or operator. `Message` history is maintained as the group changes (ie new members get to see old messages). All messages are signed by all current members so history can't be falsified unless all members decide to do so jointly.
+
+Maintaining the message history and tight linking between messages and group requires the maintenance of an index of all messages. Doing so globally would make the posting of messages racy between group members. To avoid this, the index is sharded by member in `MessageIndex`.
+
+The message history per group grows indefinitely and changes in membership require all messages to be re-signed by the new mambership. Transactions changing group membership therefore grow in complexity O(#messages * #members).
+
+## Workflow
+1. Anyone can start a group by creating a `GroupChat` contract.
+2. New group members are onboarded by existing members. Discovery happens off-ledger. Ie `Alice` is assumed to have told `Bob` and `Charlie` about her group "troll-bot" in the included scenarios.
```

Comment:
Discovery happens via `GroupMembershipRequest` -> `Onboard`, no? we can't (yet) disclose contracts off-ledger, can we?

Reply:
Ah, I see you mean discovery of the _group name_, not the cid. Disregard

Reply:
Off-ledger disclosure would be a great tool to use for stuff like this. The member that onboards the new member could disclose everything upfront, providing something like a preview of a channel in Slack, and the new joiner could then join self-service.

One day!